src/recommendation/src/main/scala/RankingAdapter.scala

```diff
@@ -0,0 +1,160 @@
+    lazy val userStructType: StructType = new StructType()
+      .add(getUserCol, IntegerType)
+      .add("recommendations", ArrayType(
+        new StructType().add(getItemCol, IntegerType).add("rating", FloatType))
+      )
+
+    lazy val itemStructType: StructType = new StructType()
+      .add(getItemCol, IntegerType)
+      .add("recommendations", ArrayType(
+        new StructType().add(getUserCol, IntegerType).add("rating", FloatType))
+      )
+
+    getMode match {
+      case "allUsers" => userStructType
+      case "allItems" => itemStructType
+      case "normal"   => userStructType
+    }
+  }
+}
+
+class RankingAdapter(override val uid: String)
+  extends Estimator[RankingAdapterModel] with ComplexParamsWritable with RankingParams with Mode {
+
+  def this() = this(Identifiable.randomUID("RecommenderAdapter"))
+
+  /** @group getParam */
+  override def getUserCol: String = getRecommender.asInstanceOf[Estimator[_] with PublicALSParams].getUserCol
+
+  /** @group getParam */
+  override def getItemCol: String = getRecommender.asInstanceOf[Estimator[_] with PublicALSParams].getItemCol
+
+  /** @group getParam */
+  override def getRatingCol: String = getRecommender.asInstanceOf[Estimator[_] with PublicALSParams].getRatingCol
+
+  def fit(dataset: Dataset[_]): RankingAdapterModel = {
+    new RankingAdapterModel()
+      .setRecommenderModel(getRecommender.fit(dataset))
+      .setMode(getMode)
+      .setK(getK)
+      .setUserCol(getUserCol)
+      .setItemCol(getItemCol)
+      .setRatingCol(getRatingCol)
+      .setLabelCol(getLabelCol)
+  }
+
+  override def copy(extra: ParamMap): RankingAdapter = {
+    defaultCopy(extra)
+  }
+
+}
+
+object RankingAdapter extends ComplexParamsReadable[RankingAdapter]
+
+/**
+  * Model from train validation split.
+  *
+  * @param uid Id.
+  */
+class RankingAdapterModel private[ml](val uid: String)
+  extends Model[RankingAdapterModel] with ComplexParamsWritable with Wrappable with RankingParams with Mode {
+
+  def this() = this(Identifiable.randomUID("RankingAdapterModel"))
+
+  val recommenderModel = new TransformerParam(this, "recommenderModel", "recommenderModel", { x: Transformer => true })
+
+  def setRecommenderModel(m: Model[_]): this.type = set(recommenderModel, m)
+
+  def getRecommenderModel: Model[_] = $(recommenderModel).asInstanceOf[Model[_]]
+
+  def transform(dataset: Dataset[_]): DataFrame = {
+    transformSchema(dataset.schema)
+
+    val windowSpec = Window.partitionBy(getUserCol).orderBy(col(getRatingCol).desc, col(getItemCol))
+
+    val perUserActualItemsDF = dataset
+      .withColumn("rank", r().over(windowSpec).alias("rank"))
+      .where(col("rank") <= getK)
+      .groupBy(getUserCol)
+      .agg(col(getUserCol), collect_list(col(getItemCol)).as(getLabelCol))
+      .select(getUserCol, getLabelCol)
+
+    val recs = getMode match {
+      case "allUsers" => this.recommendForAllUsers(getK)
+      case "allItems" => this.recommendForAllItems(getK)
```

Comment:
you had mentioned that allItems did not work. Is this still the case?

Reply:
they all work.

Reply:
It's gone. Never see it used anyway.